ambry-store/src/main/java/com/github/ambry/store/DiskManager.java

```diff
@@ -570,4 +590,138 @@ private void reportUnrecognizedDirs() {
       }
     }
   }
+
+  public DiskHealthStatus getDiskHealthStatus(){
+    return diskHealthStatus;
+  }
+
+  /**
+    Manages the disk healthchecking tasks such as creating,writing, reading, and deleting a file
+   */
+  private class DiskHealthCheck{
+    private final int operationTimeoutSeconds;      //how long a read/write operation is allotted
+    private final boolean enabled;                  //enabled healthchecking
+    private final String heathcheckDir;             //filepath of where the file will be created
+
+
+    DiskHealthCheck(int operationTimeoutSeconds, boolean enabled, String heathcheckDir){
+      this.enabled = enabled;
+      this.heathcheckDir = heathcheckDir;
+      this.operationTimeoutSeconds = operationTimeoutSeconds;
+
+    }
+
+    /**
+     * Getter utility for enabled boolean
+     * @return flag to indicate disk healthchecking or not
+     */
+    public boolean isEnabled(){
+      return enabled;
+    }
+
+    /**
+     * Attempts to close the filechannel and delete the file
+     * @param fileChannel the file used for disk healthcheck
+     * @return True/False depending on the close/delete was successful
+     */
+    private boolean deleteFile(AsynchronousFileChannel fileChannel){
+      if(fileChannel.isOpen()){
+        try{
+          fileChannel.close();
+        }catch(IOException e){
+          diskHealthStatus = diskHealthStatus.DELETE_IO_EXCEPTION;
+          logger.error("Exception occurred when deleting a file for disk healthcheck", e);
+          return false;
+        }
+      }
+      return true;
+    }
+
+    /**
+     * Performing Disk HealthCheck Test in 5 steps
+     * 1. Checks if there is sufficient space on disk
+     * 2. Creates the file on the filepath
+     * 3. Writes today's date to disk
+     * 4. Read and compares what was written to disk
+     * 5. Deletes the file
+     */
+    public void diskHealthTest(){
+      if(!enabled){ return;} //ensures disk checks are enabled
+
+      //Creates directory for the file if not already there
+      File dir = new File(heathcheckDir);
+      if (!dir.exists()) dir.mkdirs();
+
+      //Creates and opens the file
+      AsynchronousFileChannel fileChannel;
+      try {
+        Path path = Paths.get(heathcheckDir,"temp");
+        fileChannel =
+            AsynchronousFileChannel.open(path, StandardOpenOption.WRITE, StandardOpenOption.READ,
+                StandardOpenOption.CREATE, StandardOpenOption.DELETE_ON_CLOSE);
+        fileChannel.force(true); //writes to disk immediately
```

Comment:
I don't think there is anything to force at this moment, we can remove this line.

Reply:
This is meant for the future write/read calls to be written to the storage device immediately as to avoid staying in the filechannel's buffer. At least that's what I understood from 
'Forces any updates to this channel's file to be written to the storage device that contains it.'

Reply:
What you are mentioning is the DIRECT_IO mode. The open call should have a flag to open the file in that mode. Is there a flag ?

Reply:
Yeah this can be done automatically with the flag like @snalli suggested, I included the `DSYNC` flag which will synchronize with disk faster than `SYNC` as `DSYNC` doesn't need the metadata to write the file's updates to disk since we don't check for the metadata in the healthcheck.

Reply:
This filesystem in this case will sync metadata since you are creating file and allocating block. Just syncing data would lead to crash inconsistency. The FS would sync data only when you are overwriting file pages without extending or shrinking the file. Just use SYNC in your code. It would only be a few millisecs slower than DSYNC, which is tolerable given the frequency of health check is in the order of secs.

Reply:
If disk perf is a concern, then just buffer the writes in memory and flush just before closing the file. 
From what I understand, force(true) is the same as DSYNC ?